### PR TITLE
add separate game state

### PR DESCRIPTION
**Pull request by electrovir**

### CODE REVIEW

File: frontend/modules/game-state/game-action.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import {Minion} from "../minion";
```
Comment (lastmjs) [inline frontend/modules/game-state/game-action.ts:1]: I think it's best to keep the directory structure as flat as possible, we should keep all modules at the same level until absolutely necessary to nest

File: frontend/modules/game-state/game-action.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import {Player} from "../player";
+
+export type GameAction = AddMinionAction | AddPlayerAction | CreateUserAction;
+
+export type AddMinionAction = {
```
Comment (lastmjs) [inline frontend/modules/game-state/game-action.ts:6]: Action types should have the same name as their type property's value. ADD_MINION_ACTION should be the name of the type, the type of the type property, and the value of the type property

File: frontend/modules/game-state/game-action.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    type: 'add-minion',
+    minion: Minion,
+};
+
+export type AddPlayerAction = {
+    type: 'add-player',
```
Comment (lastmjs) [inline frontend/modules/game-state/game-action.ts:12]: These should all be readonly

File: frontend/modules/game-state/game-action.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import {Minion} from "../minion";
+import {Player} from "../player";
```
Comment (lastmjs) [inline frontend/modules/game-state/game-action.ts:2]: This is just a types file. It should be in the types directory

File: frontend/modules/game-state/game-store.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import {Player, generateRandomUserPlayer} from "../player";
+import {Minion} from "../minion";
+import {createObjectStore} from "reduxular";
+import {GameAction} from "./game-action";
+
+export type GameState = {
+    readonly minions: ReadonlyArray<Minion>;
+    readonly user: Readonly<Player> | undefined;
+    readonly players: {[playerId: string]: Player};
```
Comment (lastmjs) [inline frontend/modules/game-state/game-store.ts:9]: Make players its own type

File: frontend/modules/game-state/game-store.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        case 'add-minion':
+            return {
+                ...state,
+                minions: [
+                    ...state.minions,
+                    action.minion
+                ],
+            };
+        case 'add-player':
+            if (state.players.hasOwnProperty(action.player.id)) {
```
Comment (lastmjs) [inline frontend/modules/game-state/game-store.ts:31]: Why do this and not `state.players[action.player.id]`?

File: frontend/modules/game-state/game-store.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                players: {
+                    ...state.players,
+                    [action.player.id]: action.player
+                },
+            };
+        case 'create-user':
+            const user = generateRandomUserPlayer();
+            return {
+                ...state,
+                user: user,
```
Comment (lastmjs) [inline frontend/modules/game-state/game-store.ts:45]: No need to repeat user here

File: frontend/modules/minion.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export type Minion = {
```
Comment (lastmjs) [inline frontend/modules/minion.ts:1]: This should be in the types directory

File: frontend/modules/player.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import {getRandomString} from "./string";
+import {generateRandomColorString} from "./color";
+
+export type Player = {
```
Comment (lastmjs) [inline frontend/modules/player.ts:4]: Should be in the types directory

File: frontend/modules/game-state/game-action.ts (hunk 2 above)
Comment (electrovir) [reply frontend/modules/game-state/game-action.ts:6]: Changing type values to be UPPER_CASE. Leaving Action type names.

**Pull request merged**